### DISCUSSION TITLE
[apimanui] external keycloak not redirect http to https

### DISCUSSION THREAD

**Discussion by inspirerit** [apiman/apiman#1754] [Help]
I using latest version 2.0.2.Final
Flowing document [here](https://www.apiman.io/latest/production-guide.html#_keycloak) and [here](https://www.apiman.io/apiman-docs/installation-guide/2.2.0.Final/manager/production-manager.html) for setup

I using external keycloak for apimanui login
Keycloak version: 15.1.1
realm import from [here](https://github.com/apiman/apiman/blob/master/distro/data/src/main/resources/data/apiman-realm-for-keycloak.json)
1. Change **auth-server-url** and **secret** on [standalone-apiman.xml](https://github.com/apiman/apiman/blob/master/distro/wildfly/src/main/resources/overlay/standalone/configuration/standalone-apiman.xml) to point external keycloak
> 
    <secure-deployment name="apiman.war">
        <realm>${env.APIMAN_AUTH_REALM:apiman}</realm>
        <auth-server-url>https://kc-auth.xx.xx/auth</auth-server-url>
        <resource>apiman</resource>
        <credential name="secret">password</credential>
        <disable-trust-manager>true</disable-trust-manager>
        <bearer-only>true</bearer-only>
        <enable-basic-auth>true</enable-basic-auth>
        <principal-attribute>preferred_username</principal-attribute>
    </secure-deployment>
    <secure-deployment name="apimanui.war">
        <realm>${env.APIMAN_AUTH_REALM:apiman}</realm>
        <auth-server-url>https://kc-auth.xx.xx/auth</auth-server-url>
        <resource>apimanui</resource>   
        <credential name="secret">password</credential>
        <disable-trust-manager>true</disable-trust-manager>
        <principal-attribute>preferred_username</principal-attribute>
    </secure-deployment>
2. On keycloak server
- add PROXY_ADDRESS_FORWARDING = true

- configure the Valid Redirect URIs 
> 
apimanui
https://myapiman.xx.xx/apimanui/*

apiman
https://myapiman.xx.xx/apiman/*

Then,
When I access https://myapiman.xx.xx/apimanui It rediect to https://kc-auth.xx.xx/auth/realms/apiman/xxx... 
then I see **redirect_uri**= http://myapiman.xx.xx/apimanui/ (without s) and response is **invalid_redirect_uri**

If I change all setup to http, it work like a charm, but I need use SSL, so @msavy can you help me for this issue

**Comment by msavy** [comment]
Hello!

The latest version of Apiman is 2.2.0.Final - you can find links on apiman.io or the downloads section here on GitHub (and DockerHub, etc).

To fix your redirect issue:
- log into Keycloak ([your keycloak pod]/auth)
- find the right client (apimanui client)
- set the allow redirect URLs it include your https redirect (for your specific URL or pattern, however you want)
- save

See more on redirect URLs in the Keycloak documentation (https://www.keycloak.org/)

>  so @msavy can you help me for this issue

Have we met before 😅?

**Reply by inspirerit** [reply]
Have you read about my config ?
I already config valid url on keycloak server like below
> 
apimanui client:
https://myapiman.xx.xx/apimanui/*

apiman client
https://myapiman.xx.xx/apiman/*

But when login to keycloak it still redirect to **http** not **https**

https://myapiman.xx.xx/apimanui >> http://myapiman.xx.xx/apimanui

**Reply by msavy** [reply]
To confirm:
- you start on `https:// <apiman server>/apimanui`
- it redirects you to `https:// <keycloak server>`
- after you log in it redirects you back to `http://<apiman server>/apimanui` and get `invalid_redirect_uri`

**Reply by inspirerit** [reply]
Yes this flow as you said

**Reply by msavy** [reply]
Seems to work OK for me:

https://user-images.githubusercontent.com/423513/150127111-1c22fce2-f2a3-45a7-bd48-178744e1bacc.mov

**Reply by inspirerit** [reply]
Can you share your config with external keycloak ?
And If have issue about redirect https to http on keycloak, Which one(keycloak, apimanui,...) do I need to dig to find the error?

**Reply by msavy** [reply]
Something along these lines.

https://github.com/apiman/apiman/blob/3.x/manager/ui/war/ci/docker-compose.yml

**Reply by msavy** [reply]
Have you been able to resolve this?

If so, please do share your solution in case another member of the community encounters the same problem.
